internal/security/fileprovider/provider_test.go

```diff
@@ -406,3 +367,113 @@ func (wcb *writeCloserBuffer) Chmod(mode os.FileMode) error {
+
+		count++
+	}
+	jsonStr += `}`
+
+	mockFileIoPerformer.On("OpenFileReader", configFile, os.O_RDONLY, os.FileMode(0400)).Return(
+		strings.NewReader(jsonStr), nil)
+
+	for service := range expectedTokenConfs {
+		expectedSrvDir := filepath.Join(outputDir, service)
+		expectedSrvFile := filepath.Join(expectedSrvDir, outputFilename)
+		expectedSrvBuf := new(bytes.Buffer)
+		expectedEnvServiceBufMap[service] = expectedSrvBuf
+
+		mockFileIoPerformer.On("MkdirAll", expectedSrvDir, os.FileMode(0700)).Return(nil)
+		mockFileIoPerformer.On("OpenFileWriter", expectedSrvFile, os.O_CREATE|os.O_TRUNC|os.O_WRONLY,
+			os.FileMode(0600)).Return(&writeCloserBuffer{expectedSrvBuf}, nil)
+	}
+
+	mockAuthTokenLoader := &MockAuthTokenLoader{}
+	mockAuthTokenLoader.On("Load", privilegedTokenPath).Return("fake-priv-token", nil)
+
+	expectedService1Policy := `{"path":{"secret/edgex/` + serviceName + `/*":{"capabilities":["create","update","delete","list","read"]}}}`
+	expectedService1Parameters := makeDefaultTokenParameters(serviceName)
+	expectedService1Parameters["meta"] = makeMetaServiceName(serviceName)["meta"]
+	mockSecretStoreClient := &MockSecretStoreClient{}
+	mockSecretStoreClient.On("InstallPolicy", "fake-priv-token", "edgex-service-"+serviceName, expectedService1Policy).Return(http.StatusNoContent, nil)
+	mockSecretStoreClient.On("CreateToken", "fake-priv-token", expectedService1Parameters, mock.Anything).
+		Run(func(args mock.Arguments) {
+			setCreateTokenResponse(args.Get(2).(*interface{}))
+		}).Return(http.StatusOK, nil)
+
+	// setup expected things for addtional services from env if any
+
+	for service := range expectedTokenConfs {
+		expectedServicePolicy := `{"path":{"secret/edgex/` + service + `/*":{"capabilities":["create","update","delete","list","read"]}}}`
+		expectedServiceParameters := makeDefaultTokenParameters(service)
+
+		expectedServiceParameters["meta"] = makeMetaServiceName(service)["meta"]
+
+		mockSecretStoreClient.On("InstallPolicy", "fake-priv-token", "edgex-service-"+service, expectedServicePolicy).Return(http.StatusNoContent, nil)
+		mockSecretStoreClient.On("CreateToken", "fake-priv-token", expectedServiceParameters, mock.Anything).
+			Run(func(args mock.Arguments) {
+				setCreateTokenResponse(args.Get(2).(*interface{}))
+			}).
+			Return(http.StatusOK, nil)
+	}
+
+	p := NewTokenProvider(mockLogger, mockFileIoPerformer, mockAuthTokenLoader, mockSecretStoreClient)
+	p.SetConfiguration(secretstoreclient.SecretServiceInfo{}, config.TokenFileProviderInfo{
+		PrivilegedTokenPath: privilegedTokenPath,
+		ConfigFile:          configFile,
+		OutputDir:           outputDir,
+		OutputFilename:      outputFilename,
+	})
+
+	// Act
+	err := p.Run()
+
+	if errFromEnv != nil {
+		return errFromEnv
+	}
+
+	// Assert
+	// - {OutputDir}/myservice/{OutputFilename} w/proper contents
+	// - Correct policy for serviceName
+	// - Correct token parameters for serviceName
+	// - All other expectations met
+
+	assert.NoError(t, err)
```

Comment:
Should this be a require? i.e. does it make sense to continue the test if an error occurred?

Reply:
This comment misplaced?  But agree that TestTokenUsingDefaults() has more than one test encoded inside of it.  Each test should be its own test function.  I don't see a point in this function returning error...

Reply:
NM about "require" -- didn't know that existed.  In any case, personally I like to see what mocks weren't called when an error is returned; it is a great debugging aid.

Reply:
The reason I added the return error for this helper function is that I'd like to cover the expecting error case and also at the same time the error returning from `GetTokenConfigFromEnv` function call is covered to increase the testing coverage.

Reply:
ok